### PR TITLE
LangServe

### PR DESCRIPTION
**Pull request by eyurtsev**
Adds LangServe package

* Integrate Runnables with Fast API creating Server and a RemoteRunnable client
* Support multiple runnables for a given server
* Support sync/async/batch/abatch/stream/astream/astream_log on the client side (using async implementations on server)
* Adds validation using annotations (relying on pydantic under the hood) -- this still has some rough edges -- e.g., open api docs do NOT generate correctly at the moment
* Uses pydantic v1 namespace

Known issues: type translation code doesn't handle a lot of types (e.g., TypedDicts)

### CODE REVIEW

File: libs/langserve/pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+repository = "https://github.com/langchain-ai/langchain"
+
+[tool.poetry.dependencies]
+python = ">3.8.1,<4"
+httpx = "^0.25.0"
+langchain = { git = "https://github.com/langchain-ai/langchain", subdirectory = "libs/langchain" }
+fastapi = {version = ">0.90.1", optional = true}
+sse-starlette = {version = "^1.6.5", optional = true}
+httpx-sse = {version = "^0.3.1", optional = true}
+pydantic = "^1"
```
Comment (baskaryan) [inline libs/langserve/pyproject.toml:17]: why pin

File: libs/__init__.py
[the remark this answers is not in the document]
Comment (eyurtsev) [reply libs/__init__.py:1]: good catch -- not needed im careless as usual

File: libs/langserve/langserve/server.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    Union,
+)
+
+from langchain.load.dump import dumpd, dumps
+from langchain.load.load import load
+from langchain.schema.runnable import Runnable
+from typing_extensions import Annotated
+
+try:
+    from pydantic.v1 import BaseModel
```
[the remark this answers is not in the document]
Comment (eyurtsev) [reply libs/langserve/langserve/server.py:18]: Yeah -- I don't think we should couple the namespaces that way

File: libs/langserve/langserve/client.py [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+        **kwargs: Optional[Any],
+    ) -> List[Output]:
+        if not inputs:
+            return []
+        if return_exceptions:
+            raise NotImplementedError(
+                "return_exceptions is not supported for remote clients"
+            )
+
+        if isinstance(config, list):
```
[the remark this answers is not in the document]
Comment (eyurtsev) [reply libs/langserve/langserve/client.py:175]: I think better not to mix collection and element handling in one function signature, can make another standalone function -- but it's only 2 instances here so chose not do that

File: libs/langserve/langserve/server.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    create_batch_request_model,
+    create_invoke_request_model,
+    create_runnable_config_model,
+    create_stream_log_request_model,
+    create_stream_request_model,
+    replace_lc_object_types,
+)
+
+try:
+    from fastapi import APIRouter, FastAPI
```
[the remark this answers is not in the document]
Comment (eyurtsev) [reply libs/langserve/langserve/server.py:32]: Can fix up after merging -- was under the impression this is for avoiding circular deps but doesn't handle optional deps

File: libs/langserve/pyproject.toml (hunk 1 above)
Comment (eyurtsev) [reply libs/langserve/pyproject.toml:17]: :shrug: will unpin in next iteration probably after setting up CI to test with both

File: libs/langserve/langserve/__init__.py [hunk 5]
```diff
@@ -0,0 +1,4 @@
+from .client import RemoteRunnable
```
Comment (baskaryan) [inline libs/langserve/langserve/__init__.py:1]: in langchain we always do absolute imports, dunno if ew care about sticking to that

**Review by baskaryan — APPROVED** [review]
looks reasonable to me but def think @nfcampos @hwchase17 could review better than i can

File: libs/langserve/langserve/__init__.py (hunk 5 above)
Comment (eyurtsev) [reply libs/langserve/langserve/__init__.py:1]: Ah good catch -- I'm used to explicit relative, but will change to absolute to be consistent

**Pull request merged**